pfl/utils/lnrtctl/lnrtctl.8
.\" $Id$
.\" %ISC_START_LICENSE%
.\" ---------------------------------------------------------------------
.\" Copyright 2011-2018, Pittsburgh Supercomputing Center
.\" All rights reserved.
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the
.\" above copyright notice and this permission notice appear in all
.\" copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
.\" WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT SHALL THE
.\" AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
.\" DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
.\" PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
.\" TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
.\" PERFORMANCE OF THIS SOFTWARE.
.\" --------------------------------------------------------------------
.\" %END_LICENSE%
.\" %PFL_MODULES ctl %
.Dd April 19, 2016
.Dt LNRTCTL 8
.ds volume PSC \- Administrator's Manual
.Os http://www.psc.edu/
.Sh NAME
.Nm lnrtctl
.Nd
usernland Lustre networking routing daemon runtime control
.Sh SYNOPSIS
.Nm lnrtctl
.Op Fl HI
.Op Fl p Ar paramspec
.Op Fl S Ar socket
.Op Fl s Ar showspec
.Op Ar cmd arg ...
.Sh DESCRIPTION
The
.Nm
utility controls
.Xr lnrtd 8
behavior.
.Pp
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/cmd.mdoc {
.\"	daemon => "lnrtd",
.\"	cmds => { exit => "Terminate execution." },
The supported commands are as follows:
.Bl -tag -width 3n
.It Cm exit
Terminate execution.
.El
.\" }%
.Pp
The following options are recognized:
.Bl -tag -width 3n
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/H.mdoc {
.It Fl H
Suppress headers when showing data displays.
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/I.mdoc {
.It Fl I
Display large numerical values accurately instead of rounded with
postfixes of demonination such as
.Sq G
for gibibytes,
.Sq M
for mebibytes, etc.\&
.Pq Do inhuman Dc number mode .
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/p.mdoc {
.\"	log_xr => "in\n.Xr lnrtd 8\n"
.It Fl p Ar paramspec
Query/manipulate operational parameters.
.Ar paramspec
has the following format:
.Pp
.Bd -unfilled -offset 3n
.Sm off
.Oo Ar thread-name Ns Li .\& Oc Ar param
.Op Oo Li +- Oc Li = Ar value
.Sm on
.Ed
.Pp
Some parameters may be read- or write-only.
Some support modification by the assignment operators
.Li +=
and
.Li -= .
.Pp
See
.Sx Thread Specification
for details on specifying
.Ar thread-name .
The default behavior is to apply the operation to all applicable threads.
.Pp
.Ar param
may be one of the following:
.Bl -tag -width 1n -offset 3n
.It Cm log.file
File path name where log messages will be written.
This field is write-only.
If the
.Li +=
assignment operator is used, this file will not be truncated.
.It Cm log.format
The header prepended to server log messages.
See the description of
.Ev PSC_LOG_FORMAT
in
.Xr lnrtd 8
for details on this format.
.It Cm log.level Ns Op . Ns Ar subsystem
The logging level of debug message output.
.Pp
See the description of
.Ev PSC_LOG_LEVEL_ Ns Ar subsys
in
.Xr lnrtd 8
for a list of available subsystems.
If
.Ar subsystem
is not specified, all subsystems will be accessed.
.Pp
See the description of
.Ev PSC_LOG_LEVEL
in
.Xr lnrtd 8
for recognized log level values.
.It Cm log.points
List of places in the source code (in the form
.Dq file:line-number )
where granular logging can be enabled on a case-by-case
basis.
Usually assigned to by the
.Li +=
operator.
.It Cm pool. Ns Ar name
Access the memory pool specified by
.Ar name .
The following sub-fields are available:
.Pp
.Bl -tag -compact -offset 3n -width 13n
.It Cm free
Current number of allocated entries available for use in pool.
.It Cm max
Upper bound for number of entries to which auto-sized pools can grow.
.It Cm min
Lower bound for number of entries to which auto-sized pools can shrink.
.It Cm reap
Assign a value to reap the specified number of entries.
Only applies to pools that have a reaper routine.
This field is write-only.
.It Cm thres
Threshold for unused items for auto-sized pools before items are freed.
.It Cm total
Current number of entries contained in pool.
.El
.It Cm rlim
Process resource limits.
See
.Xr getrlimit 2
or
.Xr ulimit 1
for more information.
.Pp
.Bl -column "msgqueue" -offset 3n
.It Sy Name Ta Sy Field
.It =============================
.It Cm cpu        Ta Dv RLIMIT_CPU
.It Cm csize      Ta Dv RLIMIT_CORE
.It Cm dsize      Ta Dv RLIMIT_DATA
.It Cm fsize      Ta Dv RLIMIT_FSIZE
.It Cm locks      Ta Dv RLIMIT_LOCKS
.It Cm maxproc    Ta Dv RLIMIT_NPROC
.It Cm mem        Ta Dv RLIMIT_AS
.It Cm mlock      Ta Dv RLIMIT_MEMLOCK
.It Cm msgqueue   Ta Dv RLIMIT_MSGQUEUE
.It Cm nice       Ta Dv RLIMIT_NICE
.It Cm nofile     Ta Dv RLIMIT_NOFILE
.It Cm rss        Ta Dv RLIMIT_RSS
.It Cm rtprio     Ta Dv RLIMIT_RTPRIO
.It Cm rttime     Ta Dv RLIMIT_RTTIME
.It Cm sigpndg    Ta Dv RLIMIT_SIGPENDING
.It Cm stksize    Ta Dv RLIMIT_STACK
.El
.It Cm rusage
Process resource usage information.
See
.Xr getrusage 2 .
.El
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/S.mdoc {
.\"	sock => "/var/run/lnrtd. Ns Ic %h Ns Pa .sock"
.It Fl S Ar socket
Specify an alternative socket file.
The following tokens are replaced in the file name specified:
.Pp
.Bl -tag -offset 3n -width Ds -compact
.It Cm %h
the machine hostname
.It Cm %n
the daemon executable base name, i.e.\&
.Dq lnrtd
.It Cm %%
a literal
.Sq %
character
.El
.Pp
The default is
.Pa /var/run/lnrtd. Ns Ic %h Ns Pa .sock .
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/show.mdoc {
.\"	show => {
.\"	},
.It Fl s Ar showspec
Show values.
.Ar showspec
has the following format:
.Bd -unfilled -offset 3n
.Sm off
.Ar param
.Op : Ar subspec
.Sm on
.Ed
.Pp
.Ar param
may be specified as any non-ambiguous prefix abbreviation of the
following:
.Pp
.Bl -tag -width 1n -offset 3n
.It Cm loglevels
Thread logging levels.
.Ar subspec
has the following format:
.Bd -unfilled -offset 3n
.Ar thread Ns Op , Ns Ar ...
.Ed
.Pp
See
.Sx Thread Specification
for details on specifying
.Ar thread .
If
.Ar subspec
is left unspecified, all threads will be accessed.
.It Cm threads
Daemon thread activity and statistics.
.Ar subspec
has the following format:
.Bd -unfilled -offset 3n
.Ar thread Ns Op , Ns Ar ...
.Ed
.Pp
See
.Sx Thread Specification
for details on specifying
.Ar thread .
If
.Ar subspec
is left unspecified, all threads will be accessed.
.It Cm workrqs
Work requests.
.El
.Pp
The special value
.Sq \&?
may also be specified to display a list of recognized values.
.\" }%
.El
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/thr.mdoc {
.\"	thrs => {
.\"		q{lrctlacthr}			=> qq{.Nm\nconnection acceptor/multiplexor},
.\"		q{lrctlthr}			=> qq{.Nm\nconnection processor},
.\"		q{lreqpollthr}			=> qq{Lustre\n.Fn LNetEQPoll\nissuer},
.\"		q{lrtiosthr}			=> qq{Timed\n.Tn I/O\nstats updater},
.\"		q{lrusklndplthr Ns Ar %d}	=> qq{Lustre userland socket poll},
.\"	}
.Ss Thread Specification
Options which take
.Ar thread-name
parameters may be specified by one or more of the following tokens,
separated by commas:
.Pp
.Bl -tag -compact -offset 3n -width 16n
.It Cm lrctlacthr
.Nm
connection acceptor/multiplexor
.It Cm lrctlthr
.Nm
connection processor
.It Cm lreqpollthr
Lustre
.Fn LNetEQPoll
issuer
.It Cm lrtiosthr
Timed
.Tn I/O
stats updater
.It Cm lrusklndplthr Ns Ar %d
Lustre userland socket poll
.It Cm everyone
All threads
.Pq default, where applicable
.El
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/env.mdoc {
.Sh ENVIRONMENT
.Bl -tag -width 3n
.It Ev CTL_SOCK_FILE
Override the default control socket file path.
.El
.\" }%
.Sh FILES
.Bl -tag -width Pa
.It Pa /var/run/lnrtd. Ns Ic %h Ns Pa .sock
default
.Xr lnrtd 8
control socket
.El
.Sh SEE ALSO
.Xr lnrtd 8


Changes.txt
V1.0 (capps):
Capps:
	Beginning of the code base.
Isom:
	Added reread
	Added rewrite
	Added read backwards
	Added lseek+read
	Added lseek+reread

Capps:
	Added more accurate time collection method.
	Added alignment in the on chip Cache code.
	Added change step when passing 16 Meg size file.

Capps:
	Added auto+ to purge on chip cache.

kcollins:
	replaced the lseek+read &reread test with random reads and writes

Capps:
	Replaced reverse re-read with record rewrite. This gives
	both source and destination on chip cache hits.

Capps:
	added auto+multi
	Support for multiple buffers in the iozone.

Capps:
	Removed the recursion through main().
	Cleaned up the printout when not in auto mode.
	Added support for stride-read. ( Manual mode only )

Capps:
	Cleanup so it will build for bsd4_2 ( C series machines )

Capps:
	Cleanup on frontend. Now uses getopt() and has a real parser.
	Cleanup on error handling.
	Added throughput tests.
	Eliminated page faults in the throughput tests.

Capps:
	Made all perf tests table driven.
	Add back Bill's Copyright.
	ansify, prototypes, scope limitations.

V2.1 (kcollins):
	Simplified auto mode logic.  auto test now runs from
	MEGABYTES_START to MEGABYTES_END and from RECLEN_START
	to RECLEN_END with values multiplied by MULTIPLIER 
	each iteration.  Range set to (4K...16M) for RECLEN
	and (1M...512M) for MEGABYTES.  (the physical I/O
	tests for RECLEN <4K take several hours on a 1200).

	Enlarged MAXBUFFERSIZE to 16MB for large record tests.

	Added error checking for mallocs (!-).

	Changed multibuffer code to use all of MAXBUFFERSIZE
	as a circular buffer, with the number of buffers
	varying as MAXBUFFERSIZE/reclen.  This corrects 
	problems where MAXBUFFERSIZE*MAXBUFFERS was very large.
	Also modified auto mode so that tests where
	reclen>filesize are skipped.

	Modified preadv code to vary the number of buffers as
	necessary such that they will fit in min(MAXBUFFERSIZE,filesize).
	This fixes problems where the number of buffers in
	the i/o vector exceeded the size of mainbuffer.

	Added bzero for buffer when it is first malloc'd.  This
	ensures that it is initialized before use.

	Created a script (profile.fs) that runs a series of tests
	to generate a "box" around common application variables
	such as filesize, buffer size, buffer encachement, and
	number of concurrent processes.  This is intended to serve
	as the "standard" filesystem profile.

	buffer reset to mainbuffer before each test loop

V2.3 (kcollins):

	added -F option to write to specify pathnames for throughput
	tests (allowing throughput tests to multiple filesystems).

V2.4 (capps):
	Changed preadv/pwritev to use a non-sequential access pattern.
	Changed the version number.
	Moved all user interface values to KB. This simplifies
	the user interface. (consistant scaling) and it also
	allows one to start with 512kb file. This is very important
	since the first indirect block causes a significant
	slowdown in the initial write cases.

V2.5 (capps):
	Re-structure and cleanup.

V2.6 (kcollins)
	Bug fix for the throughput tests.

V2.7 (capps):
	Added -o flag. This makes all file opens for writes
	have the O_SYNC flag set. This makes all writes go
	to disk before competion. This is useful for seeing
	what the media can do without the buffer cache helping.

V2.8 (capps):
	Added -V flag. This turns on pattern verification. If
	the user were to type:
	-V 165
	Then bit pattern 0xa5 would be placed in every byte in the
	buffer and when read back from buffer cache, or disk, 
	it will be verified to be correct. If it fails then
	the error handler will specify the byte location of the
	miscompare.

V2.9 (capps):
	Added fread/re-fread, fwrite/re-fwrite to list of tests.
	Added -E to allow the user to run pread and friends as an option.

V2.10 (capps):
	Added -R. This will generate Excel compatible files that
	can then be imported into Excel and graphed.
	Added support for 5 targets to the makefile.
	Added -M  This prints out the uname -a stuff about a machine.
	Added -O  This gives all results in operations/sec instead of KB/sec.
	More code cleanup. Update comments.

V2.11 (kcollins)
	added -A.  Auto mode with no crossover and read/write tests only
	changed default record size to 64KB (from 512 bytes)

V2.12 (capps)
	Added shared memory barrier sync for throughput mode. This
	provides much finer control over the actual timeing 
	of the children.
	Added mmap() for BSD (Convex) machines that do not 
	have System V shared memory.
	Added two ways of showing throughput results. The second
	method takes into consideration children that lag behind
	due to slow devices, and gives results that are more accurate.
	Cleanup of some tab problems in throughput results.
	Cleanup of floating point output taking to much space.
	Added -d to allow a variable delay comming out of the barrier 
	in the throughput tests.
V2.12	(kcollins)
	added declaration for create_list to make ansi c compiles work
	several fixes to some of the SPPUX 5.x make targets
	added date run to banner (hope this doesn't break your scripts $-)

V2.13	(capps)
	Added "stone walling". During throughput tests, if one process
	finishes then all others are sent a signal to tell them 
	to stop. (The parallel region has finished). This provides 
	better numbers for throughput.  
	Only bzero or fill min(reclen,CACHE_SIZE) this saves a bunch
	of paging on workstations with small memory systems.
	Fixed broken target in the makefile.
	Note: use of -d is not advised. It makes the children not run 
	in parallel.
V2.14   (capps)
	Bug fix to avoid anomaly in SPP-UX. In SPP-UX the filesystem
	code preallocates meta-data to improve initial file writes.
	The first indirect block allocation was causing a block 
	of zeros to be written syncronously. In SPP-UX the filesytem
	code preallocates zero filled blocks when the first writer
	touches a filesystem after a sync. A pool of on disk zero'd 
	blocks are created asynchronously and handed out to writers

pfl/utils/lnrtctl/lnrtctl.c
/* $Id$ */
/*
 * %ISC_START_LICENSE%
 * ---------------------------------------------------------------------
 * Copyright 2011-2018, Pittsburgh Supercomputing Center
 * All rights reserved.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the
 * above copyright notice and this permission notice appear in all
 * copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
 * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT SHALL THE
 * AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
 * PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
 * TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THIS SOFTWARE.
 * --------------------------------------------------------------------
 * %END_LICENSE%
 */

#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "pfl/cdefs.h"
#include "pfl/pfl.h"
#include "pfl/ctl.h"
#include "pfl/ctlcli.h"
#include "pfl/fmt.h"
#include "pfl/log.h"

#include "../lnrtd/lnrtd.h"

struct psc_ctlshow_ent psc_ctlshow_tab[] = {
	PSC_CTLSHOW_DEFS
};

struct psc_ctlmsg_prfmt psc_ctlmsg_prfmts[] = {
	PSC_CTLMSG_PRFMT_DEFS
};

struct psc_ctlcmd_req psc_ctlcmd_reqs[] = {
};

PFLCTL_CLI_DEFS;

const char	*progname;
const char	*daemon_name = "lnrtd";

__dead void
usage(void)
{
	fprintf(stderr,
	    "usage: %s [-HIn] [-p paramspec] [-S socket] [-s value] [cmd arg ...]\n",
	    progname);
	exit(1);
}

struct psc_ctlopt opts[] = {
	{ 'H', PCOF_FLAG, &psc_ctl_noheader },
	{ 'I', PCOF_FLAG, &psc_ctl_inhuman },
	{ 'n', PCOF_FLAG, &psc_ctl_nodns },
	{ 'p', PCOF_FUNC, psc_ctlparse_param },
	{ 's', PCOF_FUNC, psc_ctlparse_show },
};

int
main(int argc, char *argv[])
{
	pfl_init();
	progname = argv[0];
	psc_ctlcli_main(PATH_CTLSOCK, argc, argv, opts,
	    nitems(opts));
	exit(0);
}
